copy/copy.go

```diff
@@ -629,21 +661,34 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 		}
 	}
 
+	// Determine if we're allowed to modify the manifest.
+	// If we can, set to the empty string. If we can't, set to the reason why.
+	cannotModifyManifestReason := ""
+	if len(sigs) > 0 {
+		cannotModifyManifestReason = "Would invalidate signatures"
+	}
+	if destIsDigestedReference {
+		cannotModifyManifestReason = "Destination specifies a digest"
+	}
+	if options.PreserveDigests {
+		cannotModifyManifestReason = "Instructed to preserve digests"
+	}
```

Comment:
This looks like a code clone above. It could be worth breaking that into a separate function to prevent diverging in the future.

Reply:
(It seems likely, but not quite obvious, to me, that the two are going to stay in sync.)

@vrothberg did you have in mind a function with three boolean parameters? A struct with named fields, to avoid three boolean values in random order? Or something with a `copier, options, sigs` parameters?

At 2 iterations, just comments linking the two (“See also $function”) might be a low-tech approach.

Reply:
It does feel that there is a cleaner version hidden, waiting to be discovered (e.g. the “get signatures that we want to copy” code can almost certainly be consolidated between the two paths), but at least the `destIsDigestedReference` check is legitimately different — so many intermediate “correct” versions are going to be more complex, and refactoring this to be _clean and simple_ feels rather out of scope of this contributed PR.

Reply:
Agreed, let's merge and keep it in the back of our heads.

Reply:
Just to tug on this thread, #1417 is a tentative idea. I don’t think that one is actually worth it as is.

Reply:
… and #1418 for the comment pair.